### PR TITLE
[Backend] Registration page 

### PR DESCRIPTION
**Pull request by DanielItkes**
## Created a registration system for our app. 
This PR comes with a place-holder register.html file just to test functionality, will be integrated with our main registration page later on.

## Things added:
1. Get input information from the registration page using django forms.
2. Create a Django user using the 'username', 'password', 'email' inputs.
3. Create our modified User with the 'DjangoUser' we've created in step 2 and 'university', 'degree' inputs.
4. Added '/register' page in 'urls.py' for our placeholder registeration page.

**This takes are of #90 (User registration).**

### CODE REVIEW

**Review by DanWaldman — APPROVED** [review]
LGTM

**Comment by Omeramsc** [comment]
Better mention it depends on #91 and rebase on top of it instead of merging a WIP. 
anyway, do take a look in #91 , I left a comment there about some field's info that should come from the backend.

**Comment by DanielItkes** [comment]
> Better mention it depends on #91 and rebase on top of it instead of merging a WIP. anyway, do take a look in #91 , I left a comment there about some field's info that should come from the backend.

Hi I've seen what you wrote to shaked and changed it so users can only choose existing uniersities/degrees.
Also, this PR is independant from the front-end for now, all webpages are just plain simple place holders that test functionality.

**Comment by Omeramsc** [comment]
> > Better mention it depends on #91 and rebase on top of it instead of merging a WIP. anyway, do take a look in #91 , I left a comment there about some field's info that should come from the backend.
> 
> Hi I've seen what you wrote to shaked and changed it so users can only choose existing uniersities/degrees. Also, this PR is independant from the front-end for now, all webpages are just plain simple place holders that test functionality.

It does depend on it in a way, as:
1. the webpage needs specific information from the backend (refer to the comment on #91), and the backend needs to pass it to it. it won't be complete otherwise. 
2. Once this is merged, it is in production. would you want users to get to those placeholder pages? 

We **can** compromise and merge backend without frontend and the opposite, but you do need to keep what I said in mind and show it in your code.

File: AcadeMeData/forms.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    class Meta1:
+        model = DjangoUser
+        fields = ['username', 'email', 'password1', 'password2']
+
+    def save(self, commit=True):
+        user = super(UserRegistrationForm, self).save(commit=False)
+        user.email = self.cleaned_data['email']
+        if commit:
+            user.save()
+        user2 = User(user=user, university=self.cleaned_data['university'], degree=self.cleaned_data['degree'])
```
Comment (Omeramsc) [inline AcadeMeData/forms.py:20]: I don't get what is done here or why we save 2 different users

**Comment by DanielItkes** [comment]
> > > Better mention it depends on #91 and rebase on top of it instead of merging a WIP. anyway, do take a look in #91 , I left a comment there about some field's info that should come from the backend.
> > 
> > 
> > Hi I've seen what you wrote to shaked and changed it so users can only choose existing uniersities/degrees. Also, this PR is independant from the front-end for now, all webpages are just plain simple place holders that test functionality.
> 
> It does depend on it in a way, as:
> 
> 1. the webpage needs specific information from the backend (refer to the comment on [[Front-End] Register UI #91](https://github.com/redhat-beyond/AcadeMe/pull/91)), and the backend needs to pass it to it. it won't be complete otherwise.
> 2. Once this is merged, it is in production. would you want users to get to those placeholder pages?
> 
> We **can** compromise and merge backend without frontend and the opposite, but you do need to keep what I said in mind and show it in your code.

So which changes should I make?
@shakedhagag  is currently working on integrating the bootstrap code of the placeholder page into the front end page, we felt that as we are using Django's authentication system we should force the frontend on the backend and not the other way around..

File: AcadeMeData/forms.py (hunk 1 above)
Comment (DanielItkes) [reply AcadeMeData/forms.py:20]: Our 'User' model has a field that contains a 'DjangoUser', so we first create the DjangoUser and use it as input in the constructor of our user.

**Review by DeanCivlin — APPROVED** [review]
LGTM

**Review by Yarboa — DISMISSED** [review]
LGTM

**Comment by Omeramsc** [comment]
> So which changes should I make?

in #91 I asked to have dynamic values instead of static values
Those should come from the backend, in a function that renders the page.

**Comment by DanielItkes** [comment]
> > So which changes should I make?
> 
> in #91 I asked to have dynamic values instead of static values Those should come from the backend, in a function that renders the page.

I think I already did it in my latest commit, I've made the forms fucntion in forms.py only accept universities and degrees that exist in the data base. Is that what you meant?

**Comment by Yarboa** [comment]
> > > So which changes should I make?
> > 
> > 
> > in #91 I asked to have dynamic values instead of static values Those should come from the backend, in a function that renders the page.
> 
> I think I already did it in my latest commit, I've made the forms function in forms.py only accept universities and degrees that exist in the data base. Is that what you meant?

No, you should read in the business logic, AKA view, and parse the parameters when you render 
the html page, see the blog post example in UI class slides

**Comment by DanielItkes** [comment]
> > > > So which changes should I make?
> > > 
> > > 
> > > in #91 I asked to have dynamic values instead of static values Those should come from the backend, in a function that renders the page.
> > 
> > 
> > I think I already did it in my latest commit, I've made the forms function in forms.py only accept universities and degrees that exist in the data base. Is that what you meant?
> 
> No, you should read in the business logic, AKA view, and parse the parameters when you render the html page, see the blog post example in UI class slides

Can you explain why we need this?
The registration works as intended and we just need to add styling into it.

**Comment by DanielItkes** [comment]
I've encountered a problem with the registration, will work on it.

**Comment by Yarboa** [comment]
> @Yarboa @bregman-arie @Omeramsc I've made a couple of changes and made a video demonstrating how everything works. Note that since then I've made the page redirect to login once the registration was finished.
> 
>  user.registration.mp4

good work, please skip the video
You can add login logout tests, please review UI test and client fixtures.
remember no test no merge :+1:

**Comment by DanWaldman** [comment]
> > @Yarboa @bregman-arie @Omeramsc I've made a couple of changes and made a video demonstrating how everything works. Note that since then I've made the page redirect to login once the registration was finished.
> > 
> >  user.registration.mp4
> 
> good work, please skip the video
> You can add login logout tests, please review UI test and client fixtures.
> remember no test no merge :+1: 
> 

I've heard other groups have merged features without tests (and they don't even know they need tests)... just saying

**Comment by Yarboa** [comment]
> > > @Yarboa @bregman-arie @Omeramsc I've made a couple of changes and made a video demonstrating how everything works. Note that since then I've made the page redirect to login once the registration was finished.
> > > user.registration.mp4
> > 
> > 
> > good work, please skip the video
> > You can add login logout tests, please review UI test and client fixtures.
> > remember no test no merge +1
> 
> I've heard other groups have merged features without tests (and they don't even know they need tests)... just saying

No answer :), it is part of learning curve to reach the minimum needed for PRs in general.

At least one client test is mandatory, login/logout it is part of UI talk. please review and give it a try, please contact through slack if things are not working

File: AcadeMeData/test_register.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        for user_data in user_bad_example:
+            invalid = False
+            form = UserRegistrationForm(data=user_data)
+            try:
+                form.save()
+            except ValueError:
+                invalid = True
+            assert invalid
+
+    def test_sign_up_valid(self, user_example):
```
Comment (DanWaldman) [inline AcadeMeData/test_register.py:40]: @Omeramsc @Yarboa 
this is the test that failed - for some reason even though all values are legal for registration - form.is_valid() - returns false. could use some help with this.

Comment (Yarboa) [reply AcadeMeData/test_register.py:40]: Did you try creating the form and validate in the shell?

Comment (DanWaldman) [reply AcadeMeData/test_register.py:38]: @Yarboa now it is fixed.

Comment (Yarboa) [reply AcadeMeData/test_register.py:38]: Yes you had to try shell
>>> form.is_valid()
False
>>> form.errors
{'university': ['Select a valid choice. Hebrew University is not one of the available choices.'], 'degree': ['Select a valid choice. Computer Science is not one of the available choices.']}
>>>

**Comment by Yarboa** [comment]
If you had clients tests you were failing in the landing pages links
see sign_in contact_us are broken

**Comment by DanWaldman** [comment]
> If you had clients tests you were failing in the landing pages links see sign_in contact_us are broken
> 
> Please fix it or open an issue

#102 resolves contact_us, sign_in link works

**Review by Yarboa — APPROVED** [review]
LGTM

**Comment by DanielItkes** [comment]
@Omeramsc Is there anything else to fix or can we merge?

**Review by razashtar93 — APPROVED** [review]
LGTM

**Review by razashtar93 — APPROVED** [review]

**Review by bregman-arie — APPROVED** [review]

**Pull request merged**